lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -462,19 +462,26 @@ monitorStakePools
     -> NetworkLayer IO t (CardanoBlock StandardCrypto)
     -> DBLayer IO
     -> IO ()
-monitorStakePools tr gp nl db@DBLayer{..} = do
-    latestGarbageCollectionEpochRef <- mkLatestGarbageCollectionEpochRef
-    let forwardHandler = forward latestGarbageCollectionEpochRef
-    cursor <- initCursor
-    traceWith tr $ MsgStartMonitoring cursor
-    follow nl (contramap MsgFollow tr) cursor forwardHandler getHeader >>= \case
-        FollowInterrupted -> traceWith tr MsgHaltMonitoring
-        FollowFailure -> traceWith tr MsgCrashMonitoring
-        FollowRollback point -> do
-            traceWith tr $ MsgRollingBackTo point
-            liftIO . atomically $ rollbackTo point
-            monitorStakePools tr gp nl db
+monitorStakePools tr gp nl DBLayer{..} =
+    monitor =<< mkLatestGarbageCollectionEpochRef
   where
+    monitor latestGarbageCollectionEpochRef = loop
+      where
+        loop = do
+            cursor <- initCursor
+            traceWith tr $ MsgStartMonitoring cursor
+            let followTrace = contramap MsgFollow tr
+            let forwardHandler = forward latestGarbageCollectionEpochRef
+            follow nl followTrace cursor forwardHandler getHeader >>= \case
+                FollowInterrupted ->
+                    traceWith tr MsgHaltMonitoring
+                FollowFailure ->
+                    traceWith tr MsgCrashMonitoring
+                FollowRollback point -> do
+                    traceWith tr $ MsgRollingBackTo point
+                    liftIO . atomically $ rollbackTo point
+                    loop
```

Comment:
What is the rationale for this change? Is this the fix?

Reply:
> What is the rationale for this change? Is this the fix?

This is explained in the message for [the commit](https://github.com/input-output-hk/cardano-wallet/pull/2153/commits/52273ea0f97977d96c762a9d6091fcf856794a68).

> Before this change, it was possible for garbage collection to happen more than once per epoch, as any rollback would recreate the mutable reference to the latest garbage collection epoch, causing the latest garbage collection epoch value to be reset to zero.
>
> This change adjusts the reference creation mechanism so that it only happens once, on startup.

Reply:
~~How does this make it run once per epoch since the reference is created only once on startup?~~

Nevermind. That is because the `latestGarbageCollectionEpochRef` is an `IORef` that gets mutated in the `forwardHandler`.

Reply:
> Nevermind. That is because the `latestGarbageCollectionEpochRef` is an `IORef` that gets mutated in the `forwardHandler`.

That's correct.

Coincidentally, it was the integration test that helped us to discover this bug. (There are several rollbacks that occur during the integration test, each of which triggers a new garbage collection.) 

When testing on `mainnet`, we did not see this behaviour, IIRC.